sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -319,28 +323,26 @@ object ScalaReflection extends ScalaReflection {
         val arguments = params.zipWithIndex.map { case ((fieldName, fieldType), i) =>
           val Schema(dataType, nullable) = schemaFor(fieldType)
           val clsName = getClassNameFromType(fieldType)
-          val newTypePath = (s"""- field (class: "$clsName", """ +
-              s"""name: "$fieldName")""") +: walkedTypePath
+          val newPathForField = walkedTypePath.copy()
+          newPathForField.recordField(clsName, fieldName)
 
           // For tuples, we based grab the inner fields by ordinal instead of name.
+          val newPath = if (cls.getName startsWith "scala.Tuple") {
+            deserializerFor(
+              fieldType,
+              addToPathOrdinal(path, i, dataType, newPathForField),
+              newPathForField)
+          } else {
+            deserializerFor(
+              fieldType,
+              addToPath(path, fieldName, dataType, newPathForField),
+              newPathForField)
+          }
           deserializerForWithNullSafety(
-            path,
+            newPath,
             dataType,
             nullable = nullable,
```

Comment:
Doesn't this revert what you did in previous PR?

Reply:
I got review comment for previous PR as well - this is the reflection of replacing function to applied expression which ends up making `deserializerForWithNullSafety` and `expressionWithNullSafety` being same.